### PR TITLE
Deploy Grafana twice: after Logging and after Monitoring stack

### PR DESCRIPTION
**Pull request by vlvasilev**
**How to categorize this PR?**

/area logging
/kind enhancement

**What this PR does / why we need it**:
When a shoot creation is stuck at worker creation, the logs from the not joining nodes are not visible in any UI, since Grafana is deployed alongside the monitoring stack, which is deployed after successful reconciliation of the worker. With this PR we deploy the Grafana after deploying the worker but we don't wait for it to be successfully reconciled. This approach sometimes leads to exclusion of the MCM dashboard from operators Grafana because the ConfigMap containing it is deployed after the deployment of the worker. In such a case, the aggregation of all extension dashboards precedes the deployment of the MCM one. The missing dashboard will be added after the next shoot reconciliation. To avoid the absence of the MCM dashboard and all dashboards in the future which will be deployed after the worker the Grafana is deployed a second time after the monitoring stack.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
/invite @istvanballok @wyb1 
**Release note**:

```bugfix operator
Fixes a bug that prevented dashboards with node information to be shown in Grafana when a shoot was stuck in worker creation.
```

### CODE REVIEW

File: pkg/operation/botanist/monitoring.go [hunk 1]
```diff
@@ -300,6 +288,47 @@ func (b *Botanist) DeploySeedMonitoring(ctx context.Context) error {
...
+	}
+
+	if err := b.deployGrafanaCharts(ctx, common.GrafanaOperatorsRole, operatorsDashboards.String(), basicAuth, common.GrafanaOperatorsPrefix); err != nil {
+		return err
+	}
+
+	if err := b.deployGrafanaCharts(ctx, common.GrafanaUsersRole, usersDashboards.String(), basicAuthUsers, common.GrafanaUsersPrefix); err != nil {
+		return err
+	}
+	return nil
```
Comment (vpnachev) [inline pkg/operation/botanist/monitoring.go]:
```suggestion
	return b.deployGrafanaCharts(ctx, common.GrafanaUsersRole, usersDashboards.String(), basicAuthUsers, common.GrafanaUsersPrefix)
```

File: pkg/operation/botanist/monitoring.go [hunk 2]
```diff
@@ -300,6 +288,47 @@ func (b *Botanist) DeploySeedMonitoring(ctx context.Context) error {
 	return nil
 }
 
+func (b *Botanist) DeploySeedGrafana(ctx context.Context) error {
```
Comment (vpnachev) [inline pkg/operation/botanist/monitoring.go:291]: Most of the code of this method is duplicated in `DeploySeedMonitoring`. 
Have you tried to call `DeploySeedGrafana` from `DeploySeedMonitoring` based on a function parameter, e.g. `withGrafana bool`?

**Review by rfranzke — COMMENTED** [review]
> In such a case, the aggregation of all extension dashboards precedes the deployment of the MCM one. The missing dashboard will be added after the next shoot reconciliation. To avoid the absence of the MCM dashboard and all dashboards in the future which will be deployed after the worker the Grafana is deployed a second time after the monitoring stack.

TBH, I have some doubts about this fact. It can easily lead to confusion and increases operational complexity. Maybe we could redeploy Grafana again at the end of the shoot flow to ensure all dashboards are present?

**Comment by vlvasilev** [comment]
> Maybe we could redeploy Grafana again at the end of the shoot flow to ensure all dashboards are present?

This is good Idea, but as as second time or as third?

**Comment by rfranzke** [comment]
> but as as second time or as third?

I didn't quite get this, what do you mean? :)

**Comment by vlvasilev** [comment]
> I didn't quite get this, what do you mean? :)

Just joking :). I know that I have to move the second deployment of the Grafana at the end.

File: pkg/operation/botanist/monitoring.go (hunk 2 above)
Comment (vlvasilev) [reply pkg/operation/botanist/monitoring.go:291]: No

**Comment by vlvasilev** [comment]
BTW, @rfranzke I didn't see a better place for the second Grafana deployment. It looks like deploying it alongside the Monitoring t is the best place for now.
@vpnachev suggested deploying the first Grafana right after the logging, so I move it.

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go [hunk 3]
```diff
@@ -369,6 +369,11 @@ func (c *Controller) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 			Fn:           flow.TaskFn(botanist.DeployWorker).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(deployCloudProviderSecret, deployReferencedResources, waitUntilInfrastructureReady, initializeShootClients, waitUntilOperatingSystemConfigReady, waitUntilNetworkIsReady),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Deploying shoot Grafana in Seed before the shoot worker",
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot_control_reconcile.go]:
```suggestion
			Name:         "Reconciling Grafana for Shoot in Seed for the logging stack",
```

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go [hunk 4]
```diff
@@ -426,6 +431,11 @@ func (c *Controller) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 			Fn:           flow.TaskFn(botanist.DeploySeedMonitoring).RetryUntilTimeout(defaultInterval, 2*time.Minute),
 			Dependencies: flow.NewTaskIDs(initializeShootClients, waitUntilTunnelConnectionExists, waitUntilWorkerReady).InsertIf(!staticNodesCIDR, waitUntilInfrastructureReady),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Deploying shoot Grafana in Seed after the shoot monitoring stack",
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot_control_reconcile.go]:
```suggestion
			Name:         "Reconciling Grafana for Shoot in Seed for the monitoring stack",
```

**Review by rfranzke — DISMISSED** [review]
Thanks!

/lgtm
/needs second-opinion (e.g. from @vpnachev 😀)

File: charts/seed-monitoring/charts/grafana/templates/grafana-deployment.yaml [hunk 5]
```diff
@@ -29,10 +29,6 @@ spec:
         role: {{ .Values.role }}
         networking.gardener.cloud/to-loki: allowed
     spec:
-      initContainers:
-      - name: init-prometheus
-        image: {{ index .Values.images "busybox" }}
-        command: ['sh', '-c', 'until wget -T 5 -qO- http://prometheus-web/-/healthy > /dev/null; do echo waiting for Prometheus; sleep 2; done;']
```
Comment (vpnachev) [inline charts/seed-monitoring/charts/grafana/templates/grafana-deployment.yaml:35]: I don't know the full details about this dependency between grafana and prometheus, but understand that without removing it you still cannot access the logs from the loki as grafana simply cannot start-up.

It will be nice to have some feedback from @wyb1 and @istvanballok on this change. 

But while you are decoupling grafana from prometheus, please reduce the `failureThreshold` of the readiness and liveness probes of prometheus here https://github.com/gardener/gardener/blob/19f46386468b601f3fb9edea74c0cd893d788c53/charts/seed-monitoring/charts/core/charts/prometheus/templates/prometheus.yaml#L87-L104

I would expect this threshold to have value between 1 and 3 usually, not 60 and 120.

**Review by vpnachev — COMMENTED** [review]
Looks good to me, just fix the log messages as suggested by @rfranzke.

File: charts/seed-monitoring/charts/grafana/templates/grafana-deployment.yaml (hunk 5 above)
Comment (vlvasilev) [reply charts/seed-monitoring/charts/grafana/templates/grafana-deployment.yaml:35]: I guess you are right for the `livenessProbe` but for `readinessProbe`, I would not agree.
If we like to give more time to Prometheuse to bring up we should increase the time of the `readinessProbe` before it fails. On the other hand, we wish to mark it as ready as soon as possible so we don't want to make a probe every 200 seconds.

Comment (vpnachev) [reply charts/seed-monitoring/charts/grafana/templates/grafana-deployment.yaml:35]: `failureThreshold` is not measured in time but in occurrences. Currently the readiness probe has to fail 120 times in a row with 5 seconds interval between each probe to finally consider the pod as not ready. In result, the pod may receive traffic for 10 full minutes when it is actually not able to accept it. And the liveness probe having `failureThreshold=60` will most probably restart the container before even to k8s to consider the pod as not ready.

As prometheus is starting slowly, it can use `initialDelaySeconds` to delay the probes.

Now I think it was not good idea to ask you to change it, probably it deserves own PR.

Comment (wyb1) [reply charts/seed-monitoring/charts/grafana/templates/grafana-deployment.yaml:35]: Without Prometheus all of the grafana dashboards that use Prometheus as a datasource would be empty. This is why it (the init container) was originally introduced since grafana is pretty useless without the data, so we wait for prometheus. Since there are some other use cases it can be safely removed. If Prometheus is not up the only consequence is that the dashboards using Prometheus as a datasource will be empty.

File: charts/seed-monitoring/charts/core/charts/prometheus/templates/prometheus.yaml [hunk 6]
```diff
@@ -89,8 +89,8 @@ spec:
             path: /-/healthy
             port: {{ .Values.port }}
             scheme: HTTP
-          failureThreshold: 60
-          periodSeconds: 5
+          failureThreshold: 3
+          periodSeconds: 100
```
Comment (vpnachev) [inline charts/seed-monitoring/charts/core/charts/prometheus/templates/prometheus.yaml]: Please, revert this change. Now, I think it is better to tackle it via separate PR.

File: charts/seed-monitoring/charts/grafana/values.yaml [hunk 7]
```diff
@@ -1,6 +1,5 @@
 images:
   grafana: image-repository:image-tag
-  busybox: image-repository:image-tag
```
Comment (vpnachev) [inline charts/seed-monitoring/charts/grafana/values.yaml:3]: I don't see this image to be used anywhere else, so you can remove it from https://github.com/gardener/gardener/blob/7ee88e716cbed49dac40d1b12336c8c998543b39/charts/images.yaml#L191-L194

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: charts/images.yaml [hunk 8]
```diff
@@ -188,10 +188,6 @@ images:
   tag: "0.9.0"
 
 # Miscellaenous
-- name: busybox
```
Comment (vpnachev) [inline charts/images.yaml:191]: You have to run `make generate` to remove it also from `charts/images.go`

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: charts/images.yaml (hunk 8 above)
Comment (vpnachev) [reply charts/images.yaml:191]: Hm, actually `make generate` does not work. Please add `./charts/...` here https://github.com/gardener/gardener/blob/6cdb8dc690c2464558c228105852eb3ac7c2c05a/Makefile#L212 and then run it.

Comment (vlvasilev) [reply charts/images.yaml:191]: mmmmm NO. It didn't work for me so I have to remove it directly fro the file.

**Review by vpnachev — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]

File: charts/images.yaml (hunk 8 above)
Comment (vpnachev) [reply charts/images.yaml:191]: Should be working with https://github.com/gardener/gardener/pull/4533
